src/Features/Core/Portable/ExtractClass/ExtractClassWithDialogCodeAction.cs

```diff
@@ -0,0 +1,264 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.ExtractInterface;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExtractClass
+{
+    internal class ExtractClassWithDialogCodeAction : CodeActionWithOptions
+    {
+        private readonly Document _document;
+        private readonly ISymbol? _selectedMember;
+        private readonly INamedTypeSymbol _selectedType;
+        private readonly IExtractClassOptionsService _service;
+
+        public TextSpan Span { get; }
+        public override string Title => FeaturesResources.Pull_members_up_to_new_base_class;
+
+        public ExtractClassWithDialogCodeAction(Document document, TextSpan span, IExtractClassOptionsService service, INamedTypeSymbol selectedType, ISymbol? selectedMember = null)
+        {
+            _document = document;
+            _service = service;
+            _selectedType = selectedType;
+            _selectedMember = selectedMember;
+            Span = span;
+        }
+
+        public override object? GetOptions(CancellationToken cancellationToken)
+        {
+            var extractClassService = _service ?? _document.Project.Solution.Workspace.Services.GetRequiredService<IExtractClassOptionsService>();
+            return extractClassService.GetExtractClassOptionsAsync(_document, _selectedType, _selectedMember)
+                .WaitAndGetResult_CanCallOnBackground(cancellationToken);
+        }
+
+        protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(object options, CancellationToken cancellationToken)
+        {
+            if (options is ExtractClassOptions extractClassOptions)
+            {
+                // Find the original type
+                var syntaxFacts = _document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var root = await _document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var originalTypeDeclaration = root.FindNode(Span).FirstAncestorOrSelf<SyntaxNode>(syntaxFacts.IsTypeDeclaration);
+
+                if (originalTypeDeclaration is null)
+                    throw new InvalidOperationException();
+
+                // Map the symbols we're removing to annotations
+                // so we can find them easily
+                var codeGenerator = _document.GetRequiredLanguageService<ICodeGenerationService>();
+                var symbolMapping = await AnnotatedSymbolMapping.CreateAsync(
```

Comment:
is this something new?  i don't recognize it.

Reply:
`AnnotatedSymbolMapping` was moved from `ExtractInterface` code to a shared location. It was originally just called SymbolMapping, and was a private class in there. It seems general purpose enough to move out